helix-vcs/src/git/blame.rs

```diff
@@ -0,0 +1,628 @@
+                        },
+                    )
+            },
+        )?;
+
+        Some(self.blame_for_line_inserted_removed(line, inserted_lines, removed_lines))
+    }
+
+    // this is a separate function for use in Tests
+    fn blame_for_line_inserted_removed(
+        &self,
+        line: u32,
+        inserted_lines: u32,
+        removed_lines: u32,
+    ) -> LineBlame {
+        // Because gix_blame doesn't care about stuff that is not commited, we have to "normalize" the
+        // line number to account for uncommited code.
+        //
+        // You'll notice that blame_line can be 0 when, for instance we have:
+        // - removed 0 lines
+        // - added 10 lines
+        // - cursor_line is 8
+        //
+        // So when our cursor is on the 10th added line or earlier, blame_line will be 0. This means
+        // the blame will be incorrect. But that's fine, because when the cursor_line is on some hunk,
+        // we can show to the user nothing at all. This is detected in the editor
+        let blame_line = line.saturating_sub(inserted_lines) + removed_lines;
+        let repo = self.repo.to_thread_local();
+
+        let commit = self
+            .blame
+            .get(&blame_line)
+            .and_then(|obj| repo.find_commit(*obj).ok());
+        let message = commit.as_ref().and_then(|c| c.message().ok());
+        let author = commit.as_ref().and_then(|c| c.author().ok());
+
+        LineBlame {
+            commit_hash: commit
+                .as_ref()
+                .and_then(|c| c.short_id().map(|id| id.to_string()).ok()),
+            author_name: author.map(|a| a.name.to_string()),
+            author_email: author.map(|a| a.email.to_string()),
+            commit_date: author.map(|a| a.time.format(gix::date::time::format::SHORT)),
+            commit_message: message.as_ref().map(|msg| msg.title.to_string()),
+            commit_body: message
+                .as_ref()
+                .and_then(|msg| msg.body.map(|body| body.to_string())),
+            time_ago: author
+                .map(|a| helix_stdx::date::format_relative_time(a.time.seconds, a.time.offset)),
+        }
+    }
+
+    /// Compute blame for this file
+    pub fn try_new(file: PathBuf) -> Result<Self> {
+        let thread_safe_repo =
+            open_repo(get_repo_dir(&file)?).context("failed to open git repo")?;
+        let repo = thread_safe_repo.to_thread_local();
+        let head = repo.head()?.peel_to_commit_in_place()?.id;
+
+        let traverse = gix::traverse::commit::topo::Builder::from_iters(
+            &repo.objects,
+            [head],
+            None::<Vec<gix::ObjectId>>,
+        )
+        .build()?;
+
+        let relative_path = file
+            .strip_prefix(
+                thread_safe_repo
+                    .path()
+                    .parent()
+                    .context("Could not get parent path of repository")?,
+            )
+            .unwrap_or(&file)
+            .to_str()
```

Comment:
Instead of this, there should be `gix::path::` conversion:

* Path to BString: https://docs.rs/gix-path/latest/gix_path/fn.try_into_bstr.html
* BString to unix separators: https://docs.rs/gix-path/latest/gix_path/fn.to_unix_separators_on_windows.html